### PR TITLE
thunderFX : pass to remove empty autocast regions

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/Lightning-AI/lightning-thunder/issues/1342

### CODE REVIEW

File: thunder/dynamo/utils.py [hunk 1]
```diff
@@ -512,3 +512,45 @@ def checkpoint_converter(gm: torch.fx.GraphModule, sub_gm: torch.fx.GraphModule)
...
+            nodes_to_erase.append(node)
+            nodes_to_erase.append(prev_node)
+
+        prev_node = node
+
+    # Erase the marked nodes.
+    for node in nodes_to_erase:
+        empty_autocast_removed_graph_module.graph.erase_node(node)
+
+    recompile_graph(empty_autocast_removed_graph_module)
```
Comment (IvanYashchuk) [inline thunder/dynamo/utils.py:555]: If `recompile_graph` is called after `remove_empty_autocast` in `ThunderCompiler. __call__` then it doesn't need to be here.

File: thunder/dynamo/utils.py [hunk 2]
```diff
@@ -512,3 +512,45 @@ def checkpoint_converter(gm: torch.fx.GraphModule, sub_gm: torch.fx.GraphModule)
...
+    prev_node = torch.fx.node.Node(graph_module.graph, "start_node", "call_function", lambda: None, None, None)
+    nodes_to_erase = []
+    for node in empty_autocast_removed_graph_module.graph.nodes:
+        # Due to these functions mapping regions created by context manager,
+        # previous `_enter_autocast` will always correspond with current `_exit_autocast`.
+        if (
+            prev_node.target == torch.amp.autocast_mode._enter_autocast
+            and node.target == torch.amp.autocast_mode._exit_autocast
+        ):
+            # NOTE: Order of node being appended matters.
```
Comment (IvanYashchuk) [inline thunder/dynamo/utils.py:542]: Thank you for explaining this order here!

File: thunder/dynamo/utils.py [hunk 3]
```diff
@@ -512,3 +512,45 @@ def checkpoint_converter(gm: torch.fx.GraphModule, sub_gm: torch.fx.GraphModule)
...
+
+    # Dummy init node.
+    prev_node = torch.fx.node.Node(graph_module.graph, "start_node", "call_function", lambda: None, None, None)
+    nodes_to_erase = []
+    for node in empty_autocast_removed_graph_module.graph.nodes:
+        # Due to these functions mapping regions created by context manager,
+        # previous `_enter_autocast` will always correspond with current `_exit_autocast`.
+        if (
+            prev_node.target == torch.amp.autocast_mode._enter_autocast
+            and node.target == torch.amp.autocast_mode._exit_autocast
```
Comment (IvanYashchuk) [inline thunder/dynamo/utils.py:540]: What other enter and exits may appear in the Dynamo graph?

**Review by IvanYashchuk — APPROVED** [review]

File: thunder/dynamo/utils.py (hunk 3 above)
Comment (kshitij12345) [reply thunder/dynamo/utils.py:540]: Besides {_enter/_exit}_autocast, from [trace_rules.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/trace_rules.py), I could find
{_enter/_exit}_dual_level for forward mode auto-diff, {_enter/_exit}_inference_mode for inference mode.

Besides these, looking at [ctx_manager.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/ctx_manager.py), there could also be  `Stream`, `Grad`, `Device` related context managers which are mapped to something like the following

```python

old_stream = torch.cuda.get_current_stream()
new_stream = torch.cuda.Stream()
set_stream(new_stream)
# Some operations
set_stream(old_stream)

```


Refs:
https://github.com/pytorch/pytorch/blob/781c68c86549aa2fd155f92afb0406b7577ab31c/torch/_dynamo/trace_rules.py#L550-L552
https://github.com/pytorch/pytorch/blob/781c68c86549aa2fd155f92afb0406b7577ab31c/torch/_dynamo/trace_rules.py#L2393-L2394

Comment (IvanYashchuk) [reply thunder/dynamo/utils.py:540]: Thank you for the links! I created an issue from this comment: https://github.com/Lightning-AI/lightning-thunder/issues/1412.

**Comment by kshitij12345** [comment]
Ping @t-vi for review and approval

**Review by t-vi — APPROVED** [review]
Thank you @kshitij12345 @IvanYashchuk

**Pull request merged**